### ISSUE TITLE
FadeInImage support for crossfading between images

### ISSUE THREAD

**Issue by btastic** [flutter/flutter#14790]
Right now it is only possible to have a FadeInImage fade out to white and then fade in to the desired loaded image. It would be great if I can just crossfade between them if I wanted to.

https://material.io/guidelines/patterns/loading-images.html#loading-images-usage

Material guidelines state that the images can also transition.

**Comment by xster** [comment]
Does something like https://docs.flutter.io/flutter/widgets/AnimatedCrossFade-class.html work?

**Comment by btastic** [comment]
I had a look at that class, but I don't know how to make it work with network images.

Right now my code looks like this:

```dart
new FadeInImage(
  placeholder: new NetworkImageWithRetry(placeholderUrl),
  image: new NetworkImageWithRetry(imageUrl),
  fit: BoxFit.fill,
)
```

I'd have to change the AnimatedCrossFade state when the image changes right?

**Comment by Herohtar** [comment]
Just came across this trying to find out if there was some way to do a cross fade with `FadeInImage`. I'd like to see it support that, as the current fade-out-then-fade-in behavior is pretty ugly, IMO.

@btastic Maybe you already found a solution, but `FadeInImage` is implemented in Dart so you can take a look at the code to see how they listen for the `NetworkImage` completion. Also, as a simpler workaround, you can put the placeholder image in a `Stack` with the `FadeInImage` and use a transparent image for the placeholder there and set the `fadeOutDuration` to `0`, kind of like this:
```dart
Stack(
  fit: StackFit.expand,
  children: <Widget>[
    Image.asset('images/defaultProfile.png'),
    FadeInImage.memoryNetwork(
      placeholder: Uint8List(512 * 512),
      image: imageUrl,
      fadeOutDuration: Duration(milliseconds: 0),
    ),
  ],
),
```

However, if you're using two network images, you're kind of doing it wrong because the whole point of `FadeInImage` is to have a local placeholder image that can be shown immediately until the network image loads...

**Comment by Hixie** [comment]
@apwilson What was the blend mode we ended up finding for doing a proper cross-fade? Do you remember?

**Comment by lukepighetti** [comment]
The fade-out and then fade-in behavior is objectively heinous. What do you guys need from me to implement cross fade as a default?

**Comment by lukepighetti** [comment]
The fade-out and then fade-in behavior is pretty heinous. Removing fadeOutDuration adds flicker and still causes fade in from a blank background. 

You can set fadeOutDuration and fadeInDuration to instant but now you don't have a FadeInImage, just a fallback image.

Finally, there's nothing wrong with using a network image for fallback when it's cached. This is a common pattern in my projects and it works beautifully.

@Hixie was there an implementation reason for the fade out and then fade in?

**Comment by flschweiger** [comment]
I also agree that the current implementation of FadeInImage does look a bit weird. Also most developers would expect a real cross-fade between the placeholder and the loaded image, not a fade out of the placeholder and then a fade in to the loaded image.

**Comment by JulianHorst** [comment]
I also agree regarding the desired outcome with FadeInImage.

In the mean time I suggest using `CachedNetworkImage` , which does exactly what "we" want here + gives us caching 🤷‍♂ :)
https://pub.dev/packages/cached_network_image

**Comment by ryaanwells** [comment]
Hey 👋 I've raised a PR to add this functionality over [here](https://github.com/flutter/flutter/pull/57223). I'm keen to open a discussion on improving this area: splitting the animation strategy (fade out/fade in, crossfade, wipe, or any animation strategy of your choosing!) could make this area of code more extensible. 

This is my first contribution so I'm not sure where the best place is for that conversation. @Hixie, would you recommend a design doc or is there a chat channel/mailing list that's more appropriate?

Cheers all!

**Comment by deakjahn** [comment]
@lukepighetti I have something similar. I don't actually use it for network loading (other packages do the whole thing automatically with caching) but I stumbled upon `FadeInImage` to use in another scenario: I only need is to display a temporary, low level version of my image for a smooth the hero animation, until it's actually replaced with the real one. I don't know what the usual experience is but Duration.zero for both durations simply results in a *weight > 0* exception for me, not simply replacing the placeholder with the final version without any glitch. Did you find any solution to seamless replacement or shall we simply use the `FadeInImage` source and delete the animation ourselves?

**Comment by adifyr** [comment]
How has this not been resolved yet? Please get this back on a milestone.
